### DISCUSSION TITLE
Applying computeJacobian and computeOffDiagJacobian

### DISCUSSION THREAD

**Discussion by andrey1887** [idaholab/moose#26107] [Q&A Modules: Solid mechanics] [answered]
Hello. Help me please to understand what difference between these two methods. I need to override it's implementation in StressDivergenceBeam but i don't entirely understand their application.

**Comment by GiudGiud** [comment]
computeJacobian is for the block-diagonal of the Jacobian while computeOffDiagJacobian is for the off-diagonal terms (coupling between different variables)
There s more information there
https://mooseframework.inl.gov/syntax/Kernels/index.html

**Reply by lindsayad** [reply]
block diagonal to be 100% clear

**Reply by andrey1887** [reply]
Thank you, @GiudGiud, @lindsayad. 
In `ComputeIncrementalBeamStrain::computeStiffnessMatrix()`  entities
```
_K11(declareProperty<RankTwoTensor>("Jacobian_11")),
_K21_cross(declareProperty<RankTwoTensor>("Jacobian_12")),
_K21(declareProperty<RankTwoTensor>("Jacobian_21")),
_K22(declareProperty<RankTwoTensor>("Jacobian_22")),
_K22_cross(declareProperty<RankTwoTensor>("Jacobian_22_cross"))
```
are computed:
for example `_K22[0] = _total_rotation[0].transpose() * K22_local * _total_rotation[0];`
i consider that as global stiffness entities for element , is it right? 

Then in `StressDivergenceBeam::computeJacobian()` `_local_ke` is filling by these elements 
So what is difference between global stiffness matrix and  `_local_ke` which "Holds local Jacobian entries" as described  ?

**Reply by andrey1887** [reply]
@GiudGiud , @lindsayad. Help me please to understand what data `_local_ke` holds.
The desription is the next:
`Holds local Jacobian entries as they are accumulated by this Kernel`
I can't understand what is `local Jacobian entries`

I think, that  `_K22[0] = _total_rotation[0].transpose() * K22_local * _total_rotation[0];`
is transformation of stiffness matrix from local to global coordinates, so  `_K22[0]` must be entity
of global stiffness matrix isn't it. Please correct me if i am not right.

**Reply by GiudGiud** [reply]
This (how components of the Jacobian are being computed) is tied to the implementation of the beam elements, and how the the stress contributes to the Jacobian
@cbolisetti knows more than me on this. 

For `local_ke` this is just some storage for the local Jacobian
This is the classic way of computing it, looping over test functions and trial functions, we do a quadrature based integration of the contribution of the kernel to the Jacobian
```
    for (_i = 0; _i < _test.size(); _i++)
      for (_j = 0; _j < _phi.size(); _j++)
        for (_qp = 0; _qp < _qrule->n_points(); _qp++)
          _local_ke(_i, _j) += _JxW[_qp] * _coord[_qp] * computeQpJac(jvar);
```

**Reply by andrey1887** [reply]
@GiudGiud, thank you for your answer.

**Reply by andrey1887** [reply]
@cbolisetti , what do think about this?

**Comment by andrey1887** [comment]
> block diagonal to be 100% clear

Is this related to `computeOffDiagJacobian` ?

**Reply by GiudGiud** [reply]
related to computeJacobian. 
As opposed to off-block-diagnoal for computeOffDiagJacobian that looks at the coupling between variables

**Reply by andrey1887** [reply]
Thank you, @GiudGiud. What are you think about Jacobian Entities: https://github.com/idaholab/moose/discussions/26107#discussioncomment-7631837

**Reply by cbolisetti** [reply, marked as answer]
I think there is some global-local confusion here. The 'global stiffness matrix' you are referring to is in the global co-ordinate system but its just for that specific element the kernel is operating on. The local in `_local_ke` does not refer to a co-ordinate system. Its only saying that it is local to this element. But it is also in the global co-ordinate system. Therefore the stiffness matrix of the element in the global co-ordinate system is fed into `_local_ke`. @GiudGiud please correct me if I am wrong.

**Reply by andrey1887** [reply]
@cbolisetti, thank you, now it's clearer for me.

**Reply by cbolisetti** [reply]
Great. Sorry for the late responses.
